lang/ir/src/ir/item_impl/tests.rs

```diff
@@ -0,0 +1,256 @@
+    assert_is_ink_impl_block_fails(
+        &syn::parse_quote! {
+            #[ink(invalid)]
+            impl MyStorage {}
+        },
+        "unknown ink! attribute (path)",
+    );
+    assert_is_ink_impl_block_fails(
+        &syn::parse_quote! {
+            #[ink(invalid)]
+            impl MyTrait for MyStorage {}
+        },
+        "unknown ink! attribute (path)",
+    );
+    assert_is_ink_impl_block_fails(
+        &syn::parse_quote! {
+            #[ink(impl)]
+            #[ink(impl)]
+            impl MyStorage {}
+        },
+        "encountered duplicate ink! attribute",
+    );
+    assert_is_ink_impl_block_fails(
+        &syn::parse_quote! {
+            #[ink(impl)]
+            #[ink(impl)]
+            impl MyTrait for MyStorage {}
+        },
+        "encountered duplicate ink! attribute",
+    );
+    assert_is_ink_impl_block_fails(
+        &syn::parse_quote! {
+            impl MyStorage {
+                #[ink(invalid)]
+                fn invalid_fn_attr(&self) {}
+            }
+        },
+        "unknown ink! attribute (path)",
+    );
+    assert_is_ink_impl_block_fails(
+        &syn::parse_quote! {
+            impl MyTrait for MyStorage {
+                #[ink(invalid)]
+                fn invalid_fn_attr(&self) {}
+            }
+        },
+        "unknown ink! attribute (path)",
+    );
+}
+
+/// Asserts that the `TryFrom` application on the given [`syn::ItemImpl`]
+/// fails with the expected error message.
+fn assert_try_from_item_impl_fails(item_impl: syn::ItemImpl, expected_err: &str) {
+    assert_eq!(
+        <ir::ItemImpl as TryFrom<syn::ItemImpl>>::try_from(item_impl)
+            .map_err(|err| err.to_string()),
+        Err(expected_err.to_string())
+    )
+}
+
+#[test]
+fn visibility_fails() {
+    assert_try_from_item_impl_fails(
+        syn::parse_quote! {
+            impl MyStorage {
+                #[ink(message)]
+                fn my_private_message(&self) {}
+            }
+        },
+        "ink! message in inherent impl blocks must have public visibility",
+    );
+    assert_try_from_item_impl_fails(
+        syn::parse_quote! {
+            impl MyStorage {
+                #[ink(constructor)]
+                fn my_private_constructor() -> Self {}
+            }
+        },
+        "ink! constructor in inherent impl blocks must have public visibility",
+    );
+    assert_try_from_item_impl_fails(
+        syn::parse_quote! {
+            impl MyTrait for MyStorage {
+                #[ink(message)]
+                pub fn my_public_message(&self) {}
+            }
+        },
+        "ink! message in trait impl blocks must have inherited visibility",
+    );
+    assert_try_from_item_impl_fails(
+        syn::parse_quote! {
+            impl MyTrait for MyStorage {
+                #[ink(constructor)]
+                pub fn my_public_constructor() -> Self {}
+            }
+        },
+        "ink! constructor in trait impl blocks must have inherited visibility",
```

Comment:
Could you explain what the issue with the code above is? In principle constructors should be public, but in this case it should be private since the property is somehow inherited?

Reply:
This is because for trait implementation blocks you cannot specify a visibility. This rule is coming from Rust.
E.g.
```rust
impl MyTrait for MyStorage {
    fn my_trait_method() {} // <-- implicitely `pub`, so cannot manually set it to `pub`
}
```